### ISSUE TITLE
UseNewtonsoft + clientless pattern, default serializer incorrectly used to serialize request body

### ISSUE THREAD

**Issue by statler** [tmenier/Flurl#783]
Using a complex JSON object - something like the following, the object is incorrectly serialized by PostJsonAsync whether using STJ or Json.Net. However, when serialized outside PostJsonAsync using default settings for Json.Net, the result is as expected.

Using latest 4pre6.

```
  var _linkItems = new
  {
      Int_Key = 0,
      Int_Box = 1111,
      Int_Link = 2222
  };
  dynamic _linkWrapper = new JObject();
  _linkWrapper["0"] = JObject.FromObject(_linkItems);
  var _params = new
  {
      pageToke = "",
      lotIntKey = 11,
      module = 2,
      lotData = new
      {
          Details = (string)null,
          Custom = (string)null,
          LinkItems = _linkWrapper,
          Security = (string)null,
          BIMLinks = (string)null
      }
  };
```

E.g this doesn't work:

```
            var path = lot_Segments.Append("MyMethod");
            var _req = _cookieSession.Request(path.ToArray());
            var _postReq = await _req
                .PostJsonAsync(_params);
```

However, this does:
```
            var json = JsonConvert.SerializeObject(_params);
            var content = new CapturedJsonContent(json);
            var _postReq = await _req
                .SendAsync(HttpMethod.Post, content);
```

The latter correctly sends the serialized string as
{"pageToke":"","lotIntKey":11,"module":2,"lotData":{"Details":null,"Custom":null,"LinkItems":{"0":{"Int_Key":0,"Int_Box":1111,"Int_Link":2222}},"Security":null,"BIMLinks":null}}

The first one mangles the content of linkitems into some strange set of [[[],[],[]][[]]] etc.

**Comment by tmenier** [comment]
What version of Flurl?

**Comment by tmenier** [comment]
I see you're using 4.0. Newtonsoft was [removed](https://github.com/tmenier/Flurl/issues/517) so JObject is no longer supported. [Neither are dynamics](https://github.com/tmenier/Flurl/issues/699). Try the just-released [Flurl.Http.Newtonsoft](https://www.nuget.org/packages/Flurl.Http.Newtonsoft) companion lib. I'm in the middle of documentation updates now, but here's the quick & dirty of how to use:

- Option 1, specific client: `flurlClient.Settings.JsonSerializer = new NewtonsoftJsonSerializer();`
- Option 2, global: `FlurlHttp.Clients.UseNewtonsoft();`

If you're already well versed in #770 (most people probably aren't), `UseNewtonsoft()` is also available on `IFlurlClientBuilder` when configuring a specific client at startup.

**Comment by statler** [comment]
Thanks Todd - this isn't just a case of STJ vs NewtonSoft. This behaviour is observed even if you are using Newtonsoft. There is something about the serialization in the PostJsonAsync that differs from doing it outside as per my workaround above.

TLDR; UseNewtonSoft does not fix the observed behaviour.

For reference - per the original report using 4pre6.

**Comment by statler** [comment]
In the client, I specified the usenewtonsoft method - travelling so can't give the exact details - but if you can't repro it is probably something not working like i expected so I will do some more research my end and get back - feel free to close it for now and I will hit it again if there is an issue.

**Comment by tmenier** [comment]
I won't keep it open forever, but I'll keep it open a bit longer if you think you'll look at it again soon-ish. ;)

All the symptoms do seem to suggest STJ is still driving the serialization, even when you think Newtonsoft is in control. I suspect it's something with how you have things wired up.

**Comment by zhzhwcn** [comment]
Same here:
`
#r "nuget: Flurl.Http.Newtonsoft, 0.9.0"
#r "nuget: Flurl.Http, 4.0.0"
using System.Net.Http;
using Flurl;
using Flurl.Http;
using Flurl.Http.Newtonsoft;
using Newtonsoft.Json.Linq;

var url = "http://bin.mailgun.net/5b58c2a";
FlurlHttp.Clients.UseNewtonsoft();
var json = new JObject();
json["test"] = "test";
json["number"] = 1;
json["date"] = DateTime.Today;
await url.PostJsonAsync(json);
`

Receives:

`
Parameter | Value |  
-- | -- | --
test | []
number | []
date | []
`

**Comment by jassent** [comment]
> Same here:
> 
> ```
> #r "nuget: Flurl.Http.Newtonsoft, 0.9.0"
> #r "nuget: Flurl.Http, 4.0.0"
> using System.Net.Http;
> using Flurl;
> using Flurl.Http;
> using Flurl.Http.Newtonsoft;
> using Newtonsoft.Json.Linq;
> 
> var url = "http://bin.mailgun.net/5b58c2a";
> FlurlHttp.Clients.UseNewtonsoft();
> var json = new JObject();
> json["test"] = "test";
> json["number"] = 1;
> json["date"] = DateTime.Today;
> await url.PostJsonAsync(json);
> ```
> 
> Receives:
> 
> Parameter	Value	 
> test	[]	
> number	[]	
> date	[]

What happens if you do this:
```
try {
    var result = await url.PostJsonAsync(json).ReceiveString();
    Console.WriteLine(result);
}
catch (Exception ex) {
  Console.WriteLine($"{ex}");
}
```
I have run into empty results due to threading and serializer exceptions. The above would help diagnose.

**Comment by zhzhwcn** [comment]
> > Same here:
> > ```
> > #r "nuget: Flurl.Http.Newtonsoft, 0.9.0"
> > #r "nuget: Flurl.Http, 4.0.0"
> > using System.Net.Http;
> > using Flurl;
> > using Flurl.Http;
> > using Flurl.Http.Newtonsoft;
> > using Newtonsoft.Json.Linq;
> > 
> > var url = "http://bin.mailgun.net/5b58c2a";
> > FlurlHttp.Clients.UseNewtonsoft();
> > var json = new JObject();
> > json["test"] = "test";
> > json["number"] = 1;
> > json["date"] = DateTime.Today;
> > await url.PostJsonAsync(json);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Receives:
> > Parameter	Value	 
> > test	[]
> > number	[]
> > date	[]
> 
> What happens if you do this:
> 
> ```
> try {
>     var result = await url.PostJsonAsync(json).ReceiveString();
>     Console.WriteLine(result);
> }
> catch (Exception ex) {
>   Console.WriteLine($"{ex}");
> }
> ```
> 
> I have run into empty results due to threading and serializer exceptions. The above would help diagnose.

{"message":"Post received. Thanks!"} test with RoslynPad 19.1 and .net 8.0.100, no exception throws.

**Comment by jassent** [comment]
> {"message":"Post received. Thanks!"} test with RoslynPad 19.1 and .net 8.0.100, no exception throws.

zhzhwcn, this looks like a successful post. Did the url receive the payload?

**Comment by astrohart** [comment]
Me too.  If I do, e.g.:

```C#
      /// <summary>
      /// Promote an account as primary account.
      /// </summary>
      Task<Response<Account>> IAccountsEndpoint.SetAccountAsPrimaryAsync(string accountId, CancellationToken cancellationToken)
      {
         return Request(AccountsEndpoint.AppendPathSegmentsRequire(accountId, "primary"))
                .PostJsonAsync(null, cancellationToken: cancellationToken)
                .ReceiveJson<Response<Account>>();
      }
```
**Listing 1.** Method under test.

And then I run the following unit test:

```C#
      [Test]
      public async Task can_set_account_as_primary()
      {
         SetupServerSingleResponse(Account3);

         var account = await client.Accounts.SetAccountAsPrimaryAsync(Account3Model.Id);

         var truth = new Response<Account>
            {
               Data = Account3Model
            };

         truth.Should().BeEquivalentTo(account);

         server.ShouldHaveCalled($"https://api.coinbase.com/v2/accounts/{Account3Model.Id}/primary")
            .WithVerb(HttpMethod.Post);
      }
```
**Listing 2.** Unit test that fails.

The failure message is:

```

Flurl.Http.FlurlParsingException : Response could not be deserialized to JSON: POST https://api.coinbase.com/v2/accounts/82de7fcd-db72-5085-8ceb-bee19303080b/primary
  ----> System.Text.Json.JsonException : The JSON value could not be converted to System.Decimal. Path: $.data.balance.amount | LineNumber: 8 | BytePositionInLine: 28.
  ----> System.InvalidOperationException : Cannot get the value of a token type 'String' as a number.
   at Flurl.Http.FlurlClient.<HandleExceptionAsync>d__28.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Flurl.Http.FlurlResponse.<GetJsonAsync>d__18`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Flurl.Http.ResponseExtensions.<ReceiveJson>d__0`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Coinbase.Tests.Endpoints.AccountTests.<can_set_account_as_primary>d__2.MoveNext() in C:\Users\Brian Hart\source\repos\bchavez\Coinbase\Source\Coinbase.Tests\Endpoints\AccountTests.cs:line 67
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at NUnit.Framework.Internal.TaskAwaitAdapter.GenericAdapter`1.GetResult()
   at NUnit.Framework.Internal.AsyncToSyncAdapter.Await[TResult](Func`1 invoke)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.RunTestMethod(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.TestMethodCommand.Execute(TestExecutionContext context)
   at NUnit.Framework.Internal.Commands.BeforeAndAfterTestCommand.<>c__DisplayClass1_0.<Execute>b__0()
   at NUnit.Framework.Internal.Commands.DelegatingTestCommand.RunTestMethodInThreadAbortSafeZone(TestExecutionContext context, Action action)
--JsonException
   at System.Text.Json.ThrowHelper.ReThrowWithPath(ReadStack& state, Utf8JsonReader& reader, Exception ex)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.JsonSerializer.ReadCore[TValue](JsonReaderState& readerState, Boolean isFinalBlock, ReadOnlySpan`1 buffer, JsonSerializerOptions options, ReadStack& state, JsonConverter converterBase)
   at System.Text.Json.JsonSerializer.ContinueDeserialize[TValue](ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack, JsonConverter converter, JsonSerializerOptions options)
   at System.Text.Json.JsonSerializer.ReadAll[TValue](Stream utf8Json, JsonTypeInfo jsonTypeInfo)
   at Flurl.Http.Configuration.DefaultJsonSerializer.Deserialize[T](Stream stream)
   at Flurl.Http.FlurlResponse.<GetJsonAsync>d__18`1.MoveNext()
--InvalidOperationException
   at System.Text.Json.Utf8JsonReader.TryGetDecimal(Decimal& value)
   at System.Text.Json.Utf8JsonReader.GetDecimal()
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.Metadata.JsonPropertyInfo`1.ReadJsonAndSetMember(Object obj, ReadStack& state, Utf8JsonReader& reader)
   at System.Text.Json.Serialization.Converters.ObjectDefaultConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, JsonSerializerOptions options, ReadStack& state)

```
**Listing 3.** Failure of the unit test.

However when I modify my method in **Listing 1** to call `ReceiveString` and then deserialize the received JSON using `Newtonsoft`, as in:

```C#
/// <summary>
/// Promote an account as primary account.
/// </summary>
async Task<Response<Account>> IAccountsEndpoint.SetAccountAsPrimaryAsync(string accountId, CancellationToken cancellationToken)
{
   using var response = Request(AccountsEndpoint.AppendPathSegmentsRequire(accountId, "primary"))
      .PostJsonAsync(null, cancellationToken: cancellationToken);
   var responseBody = await response.ReceiveString();
   if( string.IsNullOrWhiteSpace(responseBody) )
      return new Response<Account>();

   return JsonConvert.DeserializeObject<Response<Account>>(responseBody);
}
```
**Listing 4.** Method that makes my unit test pass.

I am a little confused.  It seems more intuitive to use the `ReceiveJson<Response<Account>>` call directly as in **Listing 2.**  

Am I doing something wrong in **Listing 2**?

**Comment by jassent** [comment]
@astrohart , I think your issue is different than the others in this issue. Your issue is due to how system.text.json serializes/deserializes vs Newtonsoft. With system.text.json, you have to specify how to handle things like missing values, numbers stored as strings, and setup custom converters for unusual data types like nullable booleans. Newtonsoft pre-configures a lot of this while system.text.json needs to manually configure it. The page on [migration](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/migrate-from-newtonsoft?pivots=dotnet-8-0) covers the topic in detail.

**Comment by astrohart** [comment]
@jassent Thanks for the insightful response.  I am confused...is `Flurl` using `System.Text.Json` or `Newtonsoft` as its deserializer?  Is there any reason to use one vs. the other, in principle?

**Comment by tmenier** [comment]
Sorry for the delay, this is a priority and I'll look into it as soon as I have some time. In the mean time, you could try using the serializer directly to help determine if it's a problem with that vs the registration shortcuts. It's pretty straightforward:

https://github.com/tmenier/Flurl/blob/dev/src/Flurl.Http.Newtonsoft/NewtonsoftJsonSerializer.cs

**Comment by tmenier** [comment]
Confirmed. Sorry all, better tests would have prevented this little embarrassment. In short, "global" settings as a stand-alone thing were eliminated in 4.0 but moved to a global cache of clients used just for the clientless pattern. But with that pattern, request serialization happens before a client is selected, so changing the default serializer isn't properly applying to request bodies. (It is applying to JSON responses.) I have it proven in a test and will try to get fix released in short order.

**Comment by astrohart** [comment]
Who is this "ghost" who keeps deleting my comments? Am I being a bad person or something?

**Comment by tmenier** [comment]
@astrohart This issue is bug report, not a discussion of the merits of Newtonsoft vs STJ. I've deleted several of your comments recently because they're off-topic noise. Please stop.

**Comment by astrohart** [comment]
First of all let me apologize if I've been disrespectful.  That was never my intent although I admit I was blowing off some steam. 

But there was some frustration there given I'm a contributor on a repo that uses Flurl and I had to do hours of work to use a workaround.  But it was my intention to say "me too" on this bug, until I found the workaround. 

No hard feelings?

**Comment by tmenier** [comment]
Fixed & released. The core issue was in Flurl.Http; it would have been virtually impossible to fix in Flurl.Http.Newtonsoft alone. Both packages are updated (4.0.1 and 0.9.1 respectively), please update either of them and report back here if it does not resolve the issue. Sorry again for such a blatant bug!

**Comment by amzhang** [comment]
Thank you @tmenier! You're wonderful!

**Comment by amzhang** [comment]
This might be related. I try to set global default setting in `Program.cs` to ignore null fields during JSON serialization:
```
FlurlHttp.Clients.WithDefaults(
    builder =>
        builder.WithSettings(
            settings =>
                settings.JsonSerializer = new DefaultJsonSerializer(
                    new JsonSerializerOptions
                    {
                        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
                    }
                )
        )
);
```

but it seem to have no effect. Null fields are still serialized when calling:
```
var response = await request.PostMultipartAsync(mp => {
    mp.AddJson("config", config);
});
```

Versions
```
<PackageReference Include="Flurl" Version="4.0.0" />
<PackageReference Include="Flurl.Http" Version="4.0.2" />
```

**Comment by tmenier** [comment]
@amzhang Could you open a new issue for this? Comments in closed issues don't stay on my radar, and I suspect this isn't directly related anyway.
